crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,753 @@
+                    || X509_PUBKEY_cmp(pubkey, sig->poposkInput->publicKey)
+                    || ASN1_item_verify(
+                           ASN1_ITEM_rptr(OSSL_CRMF_POPOSIGNINGKEYINPUT),
+                           sig->algorithmIdentifier, sig->signature,
+                           sig->poposkInput, X509_PUBKEY_get0(pubkey)) < 1)
+                break;
+        } else {
+            if (pubkey == NULL
+                    || req->certReq->certTemplate->subject == NULL
+                    || ASN1_item_verify(ASN1_ITEM_rptr(OSSL_CRMF_CERTREQUEST),
+                                    sig->algorithmIdentifier, sig->signature,
+                                    req->certReq,
+                                    X509_PUBKEY_get0(pubkey)) < 1)
+                break;
+        }
+        return 1;
+    case OSSL_CRMF_POPO_KEYENC:
+        /*
+         * TODO: when OSSL_CMP_certrep_new() supports encrypted certs,
+         * return 1 if the type of req->popo->value.keyEncipherment
+         * is OSSL_CRMF_POPOPRIVKEY_SUBSEQUENTMESSAGE and
+         * its value.subsequentMessage == OSSL_CRMF_SUBSEQUENTMESSAGE_ENCRCERT
+        */
+    case OSSL_CRMF_POPO_KEYAGREE:
+    default:
+        CRMFerr(CRMF_F_OSSL_CRMF_MSGS_VERIFY_POPO,
+                CRMF_R_UNSUPPORTED_POPO_METHOD);
+        return 0;
+    }
+    CRMFerr(CRMF_F_OSSL_CRMF_MSGS_VERIFY_POPO,
+            CRMF_R_UNSUPPORTED_POPO_NOT_ACCEPTED);
+    return 0;
+}
+
+/* retrieves the serialNumber of the given cert template or NULL on error */
+ASN1_INTEGER *OSSL_CRMF_CERTTEMPLATE_get0_serialNumber(OSSL_CRMF_CERTTEMPLATE *tmpl)
+{
+    return tmpl != NULL ? tmpl->serialNumber : NULL;
+}
+
+/* retrieves the issuer name of the given cert template or NULL on error */
+X509_NAME *OSSL_CRMF_CERTTEMPLATE_get0_issuer(OSSL_CRMF_CERTTEMPLATE *tmpl)
+{
+    return tmpl != NULL ? tmpl->issuer : NULL;
+}
+
+/*
+ * fill in certificate template.
+ * Any value argument that is NULL will leave the respective field unchanged.
+ */
+int OSSL_CRMF_CERTTEMPLATE_fill(OSSL_CRMF_CERTTEMPLATE *tmpl,
+                                EVP_PKEY *pubkey,
+                                const X509_NAME *subject,
+                                const X509_NAME *issuer,
+                                const ASN1_INTEGER *serial)
+{
+    if (tmpl == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_CERTTEMPLATE_FILL, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if (subject != NULL && !X509_NAME_set(&tmpl->subject, subject))
+        goto oom;
+    if (issuer != NULL && !X509_NAME_set(&tmpl->issuer, issuer))
+        goto oom;
+    if (serial != NULL) {
+        ASN1_INTEGER_free(tmpl->serialNumber);
+        if ((tmpl->serialNumber = ASN1_INTEGER_dup(serial)) == NULL)
+            goto oom;
+    }
+    if (pubkey != NULL && !X509_PUBKEY_set(&tmpl->publicKey, pubkey))
+        goto oom;
+    return 1;
+
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_CERTTEMPLATE_FILL, ERR_R_MALLOC_FAILURE);
+    return 0;
+}
+
+
+/*-
```

Comment:
this is a normal comment, thus no /*- ?

Reply:
I'd say this is a case of special formatting since each comment line is semantically one statement/sentence.
If the comment was re-formatted (with newlines elsewhere), it would become hard to parse/understand.